File: research/explorations/bench-realistic/records.py
import sys, time
from collections import namedtuple

Vec3 = namedtuple('Vec3', ['x', 'y', 'z'])

def run(n):
    s = 0
    for i in range(n):
        v = Vec3(i, i * 2, i * 3)
        d = v.x + v.y + v.z
        v2 = v._replace(x=v.x + 1)
        s += d + v2.x
    return s

N = int(sys.argv[1]) if len(sys.argv) > 1 else 1000
# warmup
for _ in range(100):
    run(N)
iters = 10000
start = time.monotonic_ns()
for _ in range(iters):
    r = run(N)
elapsed = time.monotonic_ns() - start
print(f"result:     {r}")
print(f"iterations: {iters}")
print(f"total:      {elapsed / 1e6:.2f}ms")
print(f"per call:   {elapsed // iters}ns")

File: research/explorations/bench-realistic/records.rs
use std::env;
use std::time::Instant;

#[derive(Clone)]
struct Vec3 { x: i64, y: i64, z: i64 }

fn run(n: i64) -> i64 {
    let mut s: i64 = 0;
    for i in 0..n {
        let v = Vec3 { x: i, y: i * 2, z: i * 3 };
        let d = v.x + v.y + v.z;
        let v2 = Vec3 { x: v.x + 1, ..v };
        s += d + v2.x;
    }
    s
}

fn main() {
    let n: i64 = env::args().nth(1).and_then(|s| s.parse().ok()).unwrap_or(1000);
    for _ in 0..1000 { let _ = run(n); }
    let iters: u128 = 10000;
    let start = Instant::now();
    let mut r = 0i64;
    for _ in 0..iters { r = run(n); }
    let elapsed = start.elapsed().as_nanos();
    println!("result:     {}", r);
    println!("iterations: {}", iters);
    println!("total:      {:.2}ms", elapsed as f64 / 1e6);
    println!("per call:   {}ns", elapsed / iters);
}
